File: common/src/main/java/juuxel/adorn/util/animation/Interpolator.java
package juuxel.adorn.util.animation;

import juuxel.adorn.util.Colors;
import net.minecraft.util.Mth;

@FunctionalInterface
public interface Interpolator<T> {
    Interpolator<Float> FLOAT = Mth::lerp;
    Interpolator<Double> DOUBLE = Mth::lerp;
    Interpolator<Integer> COLOR = (delta, from, to) -> {
        float alpha = FLOAT.interpolate(delta, Colors.alphaOf(from), Colors.alphaOf(to));
        float red = FLOAT.interpolate(delta, Colors.redOf(from), Colors.redOf(to));
        float green = FLOAT.interpolate(delta, Colors.greenOf(from), Colors.greenOf(to));
        float blue = FLOAT.interpolate(delta, Colors.blueOf(from), Colors.blueOf(to));
        return Colors.color(red, green, blue, alpha);
    };

    T interpolate(float delta, T from, T to);
}

File: common/src/main/kotlin/juuxel/adorn/util/animation/Interpolator.kt
package juuxel.adorn.util.animation

import juuxel.adorn.util.Colors
import juuxel.adorn.util.color
import net.minecraft.util.math.MathHelper

fun interface Interpolator<T> {
    fun interpolate(delta: Float, from: T, to: T): T

    companion object {
        val FLOAT: Interpolator<Float> = Interpolator(MathHelper::lerp)
        val DOUBLE: Interpolator<Double> = Interpolator { delta, from, to -> MathHelper.lerp(delta.toDouble(), from, to) }
        val COLOR: Interpolator<Int> = Interpolator { delta, from, to ->
            val alpha = FLOAT.interpolate(delta, Colors.alphaOf(from), Colors.alphaOf(to))
            val red = FLOAT.interpolate(delta, Colors.redOf(from), Colors.redOf(to))
            val green = FLOAT.interpolate(delta, Colors.greenOf(from), Colors.greenOf(to))
            val blue = FLOAT.interpolate(delta, Colors.blueOf(from), Colors.blueOf(to))
            color(red, green, blue, alpha)
        }
    }
}
